File: python/PhoneDataReader.py
import sys
import os.path
import struct

class PhoneDataReader:
    __dataFilePath__ = ''
    __header__ = None
    __stream__ = None
    __dataNames__ = None
    __dataCache__ = None
    def __init__(self,dataFilePath):
        self.__dataFilePath__ = dataFilePath
        self.__stream__ = open(self.__dataFilePath__,'rb')
        self.__dataNames__ = ['corp','province','city','areaCode','postCode','telecomOperator','virtualNetworkOperator','card']
        self.__dataCache__ = {}
    def getDataFilePath(self):
        return self.__dataFilePath__
    def getHeader(self):
        if self.__header__ == None :
            self.__header__ = {}
            self.__stream__.seek(0,0)
            data = self.__stream__.read(26)
            prefix,total,indexOffset,groupOffset,groupItemsCount,dataOffset,dataLength,pubDate = struct.unpack('<4siiiBiBi',data)
            self.__header__['prefix'] = prefix
            self.__header__['total'] = total
            self.__header__['indexOffset'] = indexOffset
            self.__header__['groupOffset'] = groupOffset
            self.__header__['groupItemsCount'] = groupItemsCount
            self.__header__['dataOffset'] = dataOffset
            self.__header__['dataLength'] = dataLength
            self.__header__['pubDate'] = pubDate
        return self.__header__
    def search(self,phone):
        if phone == None or isinstance(phone,str) == False or phone.isdigit() == False or len( phone) < 7 or len( phone) > 11:
            return False
        section = phone
        if len(section) > 7:
            section = section[0:7]
        section = int(section)
        header = self.getHeader()
        groupId = self.__searchGroupId__(section,header['indexOffset'],header['groupOffset'])
        if groupId != False:
            groupData = self.__getGroupDatas__(groupId)
            resut = self.__getDataByDataIdList__(groupData)
            return resut
        return False
    def __searchGroupId__(self,phone,startOffset,stopOffset):
        if startOffset <= stopOffset :
            mind = int((stopOffset - startOffset) / 8 / 2) * 8 + startOffset
            self.__stream__.seek(mind,0)
            data = self.__stream__.read(8)
            number,offset,groupId = struct.unpack('<iHH',data)
            if phone >= number and phone <= number + offset :
                return groupId
            elif phone > number :
                return self.__searchGroupId__(phone , mind + 8 , stopOffset )
            elif phone < number :
                return self.__searchGroupId__(phone , startOffset , mind - 8)
        return False
    def __getGroupDatas__(self, groupId):
        header = self.getHeader()
        result = []
        self.__stream__.seek(header["groupOffset"] + groupId * header["groupItemsCount"] * 2,0)
        for i in range(header['groupItemsCount']):
            data = self.__stream__.read(2)
            data = struct.unpack('<H',data)
            result.append(data[0])
        return result
    def __getDataByDataIdList__(self,dataIdList):
        header = self.getHeader()
        dataOffset = header['dataOffset']
        dataLength = header['dataLength']
        result = {}
        for i in range(len(dataIdList)):
            id = dataIdList[i]
            if self.__dataCache__.get(id) == None :
                self.__stream__.seek(dataOffset + id * dataLength,0)
                data = self.__stream__.read(dataLength)
                data = data.decode()
                data = data.split('\x00')[0]
                self.__dataCache__[id] = data
            result[self.__dataNames__[i]]=self.__dataCache__[id]
        return result

if __name__ == '__main__' :
    dataFilePath = os.path.abspath(__file__)
    dataFilePath = os.path.dirname(dataFilePath)
    dataFilePath = os.path.dirname(dataFilePath)
    dataFilePath += os.path.sep + 'data' + os.path.sep + 'phone.dat'
    reader = PhoneDataReader(dataFilePath)
    print(reader.search('1851105'))

File: csharp/src/PhoneDataReader/PhoneDataReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace PhoneDataReader
{
    public class PhoneDataReader : IDisposable
    {
        private readonly Dictionary<ushort, string> _dataCache = new Dictionary<ushort, string>();
        private readonly Encoding _encoding = Encoding.UTF8;
        private readonly Stream _stream;

        private PhoneDataHeader _phoneDataHeader;

        public PhoneDataReader(Stream stream)
        {
            if (stream == null)
                throw new ArgumentNullException(nameof(stream));
            _stream = stream;
        }

        public PhoneDataReader(string filename) : this(File.OpenRead(filename))
        {
        }

        public void Dispose()
        {
        }

        public PhoneDataHeader GetHeader()
        {
            if (_phoneDataHeader == null)
            {
                _phoneDataHeader = new PhoneDataHeader();
                _stream.Seek(0, SeekOrigin.Begin);
                var buffer = new byte[26];
                _stream.Read(buffer, 0, 26);
                _phoneDataHeader.Prefix = _encoding.GetString(buffer, 0, 4);
                _phoneDataHeader.Total = BitConverter.ToInt32(buffer, 4);
                _phoneDataHeader.IndexOffset = BitConverter.ToInt32(buffer, 8);
                _phoneDataHeader.GroupOffset = BitConverter.ToInt32(buffer, 12);
                _phoneDataHeader.GroupItemsCount = buffer[16];
                _phoneDataHeader.DataOffset = BitConverter.ToInt32(buffer, 17);
                _phoneDataHeader.DataLength = buffer[21];
                _phoneDataHeader.PubDate = BitConverter.ToInt32(buffer, 22);
            }
            return _phoneDataHeader;
        }

        public PhoneData Search(string phone)
        {
            if (string.IsNullOrEmpty(phone) || (phone.Length < 7) || (phone.Length > 11))
                throw new ArgumentNullException(nameof(phone));
            var phoneSection = phone;
            if (phoneSection.Length > 7)
            {
                phoneSection = phone.Substring(0, 7);
            }
            int number;
            if (int.TryParse(phoneSection, out number))
            {
                var header = GetHeader();
                var groupId = SearchGroupId(number, header.IndexOffset, header.GroupOffset);
                if (groupId != null)
                {
                    var dataIdList = GetGroupDatas(groupId.Value);
                    var data = GetDatas(dataIdList);
                    var result = new PhoneData();
                    for (var i = 0; i < data.Count; i++)
                    {
                        var value = data[i];
                        switch (i)
                        {
                            case 0:
                                result.Corp = value;
                                break;
                            case 1:
                                result.Province = value;
                                break;
                            case 2:
                                result.City = value;
                                break;
                            case 3:
                                result.AreaCode = value;
                                break;
                            case 4:
                                result.PostCode = value;
                                break;
                            case 5:
                                result.TelecomOperator = value;
                                break;
                            case 6:
                                result.VirtualNetworkOperator = value;
                                break;
                            case 7:
                                result.Card = value;
                                break;
                        }
                    }
                    return result;
                }
            }
            else
            {
                throw new ArgumentException("手机号码格式不正确", nameof(phone));
            }
            return null;
        }

        private ushort? SearchGroupId(int phone, int startOffset, int stopOffset)
        {
            if (startOffset > stopOffset)
            {
                return null;
            }
            var mind = (stopOffset - startOffset)/8/2*8 + startOffset;
            _stream.Seek(mind, SeekOrigin.Begin);
            var index = new PhoneDataIndex();
            index.Read(_stream);
            if (phone >= index.Number && phone <= index.Number + index.Offset)
            {
                return index.GroupId;
            }
            if (phone < index.Number)
            {
                return SearchGroupId(phone, startOffset, mind - 8);
            }
            if (phone > index.Number)
            {
                return SearchGroupId(phone, mind + 8, stopOffset);
            }
            return null;
        }

        private List<ushort> GetGroupDatas(ushort groupId)
        {
            var header = GetHeader();
            _stream.Seek(header.GroupOffset + groupId*header.GroupItemsCount*2, SeekOrigin.Begin);
            var length = _phoneDataHeader.GroupItemsCount;
            var list = new List<ushort>();
            var buffer = new byte[length*2];
            _stream.Read(buffer, 0, buffer.Length);
            for (var i = 0; i < length; i++)
            {
                list.Add(BitConverter.ToUInt16(buffer, i*2));
            }
            return list;
        }

        private List<string> GetDatas(List<ushort> dataIdList)
        {
            var names = new List<string>();
            var header = GetHeader();
            foreach (var id in dataIdList)
            {
                if (!_dataCache.ContainsKey(id))
                {
                    _stream.Seek(header.DataOffset + id*header.DataLength, SeekOrigin.Begin);
                    var buffer = new byte[header.DataLength];
                    _stream.Read(buffer, 0, buffer.Length);
                    var name = _encoding.GetString(buffer).Trim('\0');
                    _dataCache.Add(id, name);
                }
                names.Add(_dataCache[id]);
            }

            return names;
        }
    }
}